test/ct_test.c

```diff
@@ -0,0 +1,407 @@
+ *
+ * Author:    Rob Percival (robpercival@google.com)
+ * Date:        2016-01-26
+ *
+ * ====================================================================
+ * Copyright (c) 2016 The OpenSSL Project.    All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *        notice, this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *        notice, this list of conditions and the following disclaimer in
+ *        the documentation and/or other materials provided with the
+ *        distribution.
+ *
+ * 3. All advertising materials mentioning features or use of this
+ *        software must display the following acknowledgment:
+ *        "This product includes software developed by the OpenSSL Project
+ *        for use in the OpenSSL Toolkit. (http://www.OpenSSL.org/)"
+ *
+ * 4. The names "OpenSSL Toolkit" and "OpenSSL Project" must not be used to
+ *        endorse or promote products derived from this software without
+ *        prior written permission. For written permission, please contact
+ *        licensing@OpenSSL.org.
+ *
+ * 5. Products derived from this software may not be called "OpenSSL"
+ *        nor may "OpenSSL" appear in their names without prior written
+ *        permission of the OpenSSL Project.
+ *
+ * 6. Redistributions of any form whatsoever must retain the following
+ *        acknowledgment:
+ *        "This product includes software developed by the OpenSSL Project
+ *        for use in the OpenSSL Toolkit (http://www.OpenSSL.org/)"
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE OpenSSL PROJECT AS IS'' AND ANY
+ * EXPRESSED OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.    IN NO EVENT SHALL THE OpenSSL PROJECT OR
+ * ITS CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+ * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ * ====================================================================
+ */
+
+#include <ctype.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "crypto/include/internal/ct_int.h"
+#include "openssl/err.h"
+#include "openssl/safestack.h"
+#include "openssl/ssl.h"
+#include "openssl/x509.h"
+#include "openssl/x509v3.h"
+#include "testutil.h"
+
```

Comment:
It looks like heartbeat_test.c is doing similar things, except it can (mostly) use the public API because one exists for SSL. I'm not very familiar with OpenSSL, but I'm assuming the reason it includes ssl_locl.h is because it's also accessing some internals.

The functions under test are declared in crypto/include/internal/ct_int.h. Despite the fact that "internal" is right there in the path name, I'd argue they do present a testable API because that directory seems intended for exposing functions for use across crypto. However, rather than debate where they fall on the public-private spectrum, I think the simplest way to resolve this is to expose those functions publicly by moving their declarations from "crypto/include/internal/ct_int.h" to "include/openssl/ct.h". How does that sound? Hopefully it's possible to get that in before the final alpha release.